### PR TITLE
Data Validators - Identity Columns and Reservoir Sampling

### LINKED ISSUE #1108: Support keeping a sample of failed row ids for data validators
## Description

When running the data validation as part of profiling, we should support user specifying a row id and then we can keep a buffer of the row ids of any failed row.

- [x] I have reviewed the [Guidelines for Contributing](https://github.com/whylabs/whylogs/blob/mainline/.github/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/whylabs/whylogs/blob/mainline/.github/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by FelipeAdachi**
## Description

- Reservoir sampling:
  - Sampling failed conditions defaults to true (sampling size default is 10)
  - If identity column not passed, samples actual validated values
  - If identity column is passed, samples row ids for failed values
- Identity columns
  - Defaults to None
  - If identity column is passed, the row ids are passed to the actions, in addition to being sampled (if sampling is enabled)

### CODE REVIEW

File: python/whylogs/core/validators/condition_validator.py [hunk 1]
```diff
@@ -1,32 +1,72 @@
...
-from whylogs.core.preprocessing import PreprocessedColumn
+from whylogs.core.metrics.metrics import MetricConfig
+from whylogs.core.stubs import pd
 from whylogs.core.validators.validator import Validator
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True)
+class ConditionValidatorConfig(MetricConfig):
```
Comment (jamie256) [inline python/whylogs/core/validators/condition_validator.py]: Should we put this into metric config?

File: python/whylogs/core/validators/condition_validator.py [hunk 2]
```diff
@@ -1,32 +1,72 @@
...
+    """
+    Configuration for a condition validator.
+
+    Args:
+        validator_sample_size (int): The number of failed values to sample. Defaults to 10.
```
Comment (jamie256) [inline python/whylogs/core/validators/condition_validator.py]: If used in a datasketch k param, this wouldn't be the count of values but a log(number_of_samples) = k, need to update the description.

File: python/whylogs/core/validators/condition_validator.py [hunk 3]
```diff
@@ -1,32 +1,72 @@
...
-    def columnar_validate(self, data: PreprocessedColumn) -> None:
+    def columnar_validate(self, data: Any, sampling_size: int = 10, identity_values: Any = None) -> None:
         count = 0
         count_failures = 0
-        for x in list(chain.from_iterable(data.raw_iterator())):
+        validate_with_row_id = False
+        if not isinstance(sampling_size, int):
+            raise ValueError("sampling_size must be an integer")
+        if self.enable_sampling:
+            self._sampler = ds.var_opt_sketch(k=sampling_size)
```
Comment (jamie256) [inline python/whylogs/core/validators/condition_validator.py]: Can we move to *init__?

**Review by jamie256 — COMMENTED** [review]
Some initial comments on code walkthrough we did.

File: python/whylogs/core/validators/condition_validator.py (hunk 2 above)
Comment (jamie256) [reply python/whylogs/core/validators/condition_validator.py]: Ok, I am wrong, its actually the number of samples

File: python/whylogs/core/validators/condition_validator.py [hunk 4]
```diff
@@ -41,11 +81,30 @@ def columnar_validate(self, data: PreprocessedColumn) -> None:
...
                     for action in self.actions:
-                        action(self.name, cond_name, x)
+                        if validate_with_row_id:
+                            action(self.name, cond_name, x, identity_values[index])
+                        else:
+                            action(self.name, cond_name, x)  # type: ignore
 
         self.total = count
 
+    def sample_failed_conditions(self) -> List[Any]:
```
Comment (jamie256) [inline python/whylogs/core/validators/condition_validator.py]:
```suggestion
    def get_samples(self) -> List[Any]:
```

To match the sketch's name. We aren't actively computing the sample with this, just reading it out of the sketch.

File: python/whylogs/core/validators/condition_validator.py [hunk 5]
```diff
@@ -1,32 +1,72 @@
...
 
-    def columnar_validate(self, data: PreprocessedColumn) -> None:
+    def columnar_validate(self, data: Any, sampling_size: int = 10, identity_values: Any = None) -> None:
         count = 0
         count_failures = 0
-        for x in list(chain.from_iterable(data.raw_iterator())):
+        validate_with_row_id = False
+        if not isinstance(sampling_size, int):
+            raise ValueError("sampling_size must be an integer")
+        if self.enable_sampling:
```
Comment (jamie256) [inline python/whylogs/core/validators/condition_validator.py]:
```suggestion
        if self.enable_sampling and self._sampler is None:
```

File: python/whylogs/core/validators/condition_validator.py (hunk 1 above)
Comment (FelipeAdachi) [reply python/whylogs/core/validators/condition_validator.py]: done in latest commit

File: python/whylogs/core/validators/condition_validator.py (hunk 3 above)
Comment (FelipeAdachi) [reply python/whylogs/core/validators/condition_validator.py]: So, it was not as straightforward to simply move to init: when the validator is instantiated, it doesn't have access to the sampling size, as this was present only in the `schema`.

To be able to put this on `post_init`, I had to remove the `sample_size` param from `MetricConfig` and put it as a parameter of the `Validator` itself.

**Review by jamie256 — APPROVED** [review]
LGTM!

**Pull request merged**